### PR TITLE
Support rendering contents inside block

### PR DESCRIPTION
**Pull request by k0kubun**
As we discussed in https://github.com/hamlit/hamlit-block/issues/1, hamlit-block has incompatibility in rendering contents inside block.
It's a kind of trade-off between the maintainability of cells-hamlit (you can implement `capture` without using output buffers) and compatibility with original Hamlit.

To ease migration to cells, it'd be good to keep compatibility in block rendering. This pull request achieves it in the same way as cells-slim. The difference between this version and v0.1.0 is that we can use hamlit-cells in non-rails environment. :)

### CODE REVIEW

**Comment by k0kubun** [comment]
@apotonick @timoschilling @seuros pinging. What do you think about this?

**Comment by apotonick** [comment]
That works, but has two things I wanted to avoid.
1. We now copy that `with_output_buffer` code **again** from Rails and I wanted to avoid this by having an explicit capture mechanism. The trick with the instance variable `@output_buffer` works, but is extremely unclean and a dirty work-around for a flawed `capture` concept. I would prefer removing this once and forever and make the template engine responsible for capture/iteration support. This is because ...
2. ...I now have to copy that horrible `with_output_buffer` shit to all gems using cells-hamlit. For example, in Formular, we now need the `@output_buffer` hack instead of simply implementing `capture` with a `yield`. This is OK and I can do it, but I hate it. 

My goal is to change how capturing is handled on a fundamental tier, and not just hack on to keep things working somehow. Objections?

**Comment by timoschilling** [comment]
@apotonick I agree with you, but maybe we should have something like:

``` ruby
class Form::Cell
  include Cell::Rails::Capture
```

to support the rails way in `cells-rails`

**Comment by apotonick** [comment]
I just tested your patch, and now Formular is broken for the following reason.

Sometimes, actually, always, captured blocks are being passed to completely different object instances. When yielding there, changing `@output_buffer` doesn't work anymore, since the block is executed in the old context. This is exactly why I favor the `hamlit-block` style.

Honestly, I'm not interested in the old style `@output_buffer` approach anymore. It is an outdated concept from times where a view was one global instance. Cells don't work like that, and if we come up with a clever way how to handle iterations, many people will change their code to make it work with this new, and way cleaner approach.

I understand this must sound odd. However, I've wasted years of trying to make every template gem work the way we want it, and failed, so I'm more than happy to simply change view semantics for the sake of a way simpler code base.

**Comment by apotonick** [comment]
Here's the main problem, again, @timoschilling (and of course the one and only @k0kubun ): Capturing using the `@output_buffer` hack works, as long as the block is executed in the same _cell instance_.

This breaks, though, as soon as you pass the block into another cell (or form builder, or whatever) object and yield the block there, because the block is evaled in the old context and the new `@output_buffer`'s not referenced/used.

My intuitive response is: drop the old, global-thinking behaviour and change the way we handle iterations (as they are the only problem with the new `-block` behaviour),

**Comment by apotonick** [comment]
What we could do is allowing to include the new-school `-block` behavior per cell. That's the good part about small objects :grin: That could work and provide a nice transition.

**Comment by k0kubun** [comment]
First of all, thank you for your comments for this pull request!

> My goal is to change how capturing is handled on a fundamental tier, and not just hack on to keep things working somehow. Objections?
> I understand this must sound odd. However, I've wasted years of trying to make every template gem work the way we want it, and failed, so I'm more than happy to simply change view semantics for the sake of a way simpler code base.

No objection. I admit that rails' way is dirty and has limits in some problems. I admire your vision that makes things better.

> Sometimes, actually, always, captured blocks are being passed to completely different object instances. When yielding there, changing @output_buffer doesn't work anymore, since the block is executed in the old context. This is exactly why I favor the hamlit-block style.
> This breaks, though, as soon as you pass the block into another cell (or form builder, or whatever) object and yield the block there, because the block is evaled in the old context and the new @output_buffer's not referenced/used.

Ah now I understand my patch's defect. It's the problem I couldn't care about.

> I agree with you, but maybe we should have something like:
> to support the rails way in cells-rails
> What we could do is allowing to include the new-school -block behavior per cell. That's the good part about small objects :grin: That could work and provide a nice transition.

Then you're okay for this direction? cells-hamlit users manually introduce `hamlit-block`'s behavior to each cell and no non-cells template is affected. While I still worry that cells-hamlit users can't render Haml templates in the same way as normal templates in cells, the solution looks much better than current implementation. :)

**Comment by apotonick** [comment]
Yeah, maybe I should not include `hamlit-block` per default in `cells-hamlit`.

**Comment by jhblacklock** [comment]
The only way I was able to get this patch to work consistently in Rails was to revert to using ActiveSupport::SafeBuffer instead of the ViewModel::OutputBuffer and set html_safe to false.  Using the patch in blocks that had Rails specific ActiveSupport::SafeBuffer instances, like fields_for blocks, would throw errors.

```
no implicit conversion of ActiveSupport::SafeBuffer into Array
```

I know this is counter to @apotonick's approach to decouple things.  Just wanted to document my temporary fix if it helps anyone.

``` ruby
def with_output_buffer(block_buffer=ActiveSupport::SafeBuffer.new)
  block_buffer.instance_variable_set(:@html_safe, false)

  @output_buffer, old_buffer = block_buffer, @output_buffer
  yield
  block_buffer
ensure
  @output_buffer = old_buffer
end
```

For multi-nested forms using fields_for, I also had to disable SafeBuffer's escaping.

``` ruby
def capture(*args)
  value = nil
  buffer = with_output_buffer { value = yield(*args) }
  buffer.to_s if buffer.size > 0
  raw(value)
end
```

---

cells (4.1.1)
cells-hamlit (0.2.0) with this PR

**Comment by benhutton** [comment]
@k0kubun @apotonick has there been any progress on "`cells-hamlit` users manually introduce `hamlit-block`'s behavior to each cell and no non-cells template is affected"???

We are attempting to slowly migrate a very large codebase to use cells, and we're running into problems right away with this. Being able to enable `hamlit-block` for the cells and disable it for everything else would be amazing!

**Comment by k0kubun** [comment]
No progress. While it's not so difficult to separate these engines, I don't know how to enable it in cells regardless of ActionView's template handler. Probably it's possible but I haven't researched it yet.

@apotonick Is there any sample doing that? It seems that cells-slim doesn't do it.

**Comment by apotonick** [comment]
Hey, i was playing more with it last week. Why can't we have the same behavior that Slim exposes, because it works perfectly fine there! This is because Slim assigns every line to a **local** variable. And: Slim also supports capture natively! Should we have a Hangout next week to discuss this?

**Comment by k0kubun** [comment]
hamlit-block's engine proved that we can cleanly achieve capture with assigning every line to a local variable and I think we can implement capture natively in hamlit-block too. I'm okay to have a Hangout to discuss this.

**Comment by benhutton** [comment]
@k0kubun @apotonick any progress guys?

**Comment by apotonick** [comment]
@benhutton What do you think this is? :angry:

**Comment by benhutton** [comment]
@apotonick sorry, let me rephrase: it sounded like from your conversation last week you were planning to meet to discuss steps forward. That's what I'm wondering about... progress on that conversation and planning out a solution (or even deciding if there's going to be a solution), not progress on writing code.

Did that meeting happen? Is there a way forward here?

**Comment by apotonick** [comment]
That's better! :joy: I am actually working on making ERBse (our own ERB implementation) capture-savvy, and once this is finished, I'll have a chat with @k0kubun so we finally get things rolling.

**Comment by emptyflask** [comment]
I'd really like it if I could use hamlit both inside and outside of cells, with working blocks and `each` iterators. I'm not sure if @apotonick and @k0kubun have decided on a direction yet, but is there anything I can do to assist?

**Comment by k0kubun** [comment]
I built cleaner hamlit-block as v0.4.0 now https://rubygems.org/gems/hamlit-block/versions/0.4.0. I wrote details in https://github.com/trailblazer/cells-hamlit/issues/6. My future direction is https://github.com/trailblazer/cells-hamlit/pull/2#issuecomment-185169764. It'd be nice to have a new integration module.

But I am not an expert of cells and don't know how to integrate template engines with cells. Maybe @apotonick knows what we should do. 😇 

I hope https://github.com/trailblazer/cells-hamlit/issues/6 will be closed by @apotonick soon. Closing this PR. Happy hacking!
